Improve Span bounds check elimination by adding `_assumeNonNegative` to `count`

eeckstein opened the pull request:
rdar://156345209

stephentyrone commented:
@eeckstein let's roll a revert of the [Mutable]Span changes from https://github.com/swiftlang/swift/pull/83150 into this for the main branch.

eeckstein committed:
stdlib: revert workaround for Span bounds checking

eeckstein committed:
stdlib: specify the `count` of `Span` and `MutableSpan` as `_assumeNonNegative`

eeckstein committed:
IRGen: better code generation for the _assumeNonNegative intrinsic

So far we only supported this for `load` and `call` arguments. Now we use `llvm.assume` for other kind of argument values.
This helps optimizing bounds checking for `Span`.

eeckstein commented:
@swift-ci apple silicon benchmark
